File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/impl/utils/PolarFirmwareUpdateUtils.kt
package com.polar.sdk.impl.utils

import com.polar.androidcommunications.api.ble.BleLogger
import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpClient
import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpUtils.PftpResponseError
import com.polar.androidcommunications.http.client.HttpResponseCodes
import com.polar.androidcommunications.http.fwu.FirmwareUpdateApi
import com.polar.androidcommunications.http.fwu.FirmwareUpdateRequest
import com.polar.sdk.api.errors.PolarBleSdkInternalException
import com.polar.sdk.api.model.FirmwareUpdateStatus
import com.polar.sdk.api.model.PolarDeviceUuid
import com.polar.sdk.api.model.PolarFirmwareVersionInfo
import fi.polar.remote.representation.protobuf.Device
import fi.polar.remote.representation.protobuf.Structures
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.File
import java.util.zip.ZipInputStream

internal object PolarFirmwareUpdateUtils {

    /**
     * Comparator for sorting FW files so that the order doesn't matter as long as
     * the SYSUPDAT.IMG file is the last one (since it makes the device boot itself).
     */
    class FwFileComparator : Comparator<File> {
        override fun compare(f1: File, f2: File): Int {
            return PolarRuntimePlannerAdapter.firmwareFilePriority(f1.name).compareTo(PolarRuntimePlannerAdapter.firmwareFilePriority(f2.name))
        }
    }

    const val BUFFER_SIZE = 8192

    private val DEVICE_FIRMWARE_INFO_PATH: String
        get() = PolarRuntimePlannerAdapter.firmwareDeviceInfoPath()
    private const val TAG = "PolarFirmwareUpdateUtils"

    suspend fun readDeviceFirmwareInfo(client: BlePsFtpClient, deviceId: String): PolarFirmwareVersionInfo {
        BleLogger.d(TAG, "readDeviceFirmwareInfo: $deviceId")
        val plan = PolarRuntimePlannerAdapter.planFileFacade("firmware-read-device-info", "GET", DEVICE_FIRMWARE_INFO_PATH)
        val response = client.request(
            PolarRuntimePlannerAdapter.fileOperationBytes(plan)
        )
        val proto = Device.PbDeviceInfo.parseFrom(response.toByteArray())
        return PolarFirmwareVersionInfo(
            deviceFwVersion = devicePbVersionToString(proto.deviceVersion),
            deviceModelName = proto.modelName,
            deviceHardwareCode = proto.hardwareCode
        )
    }

    fun isAvailableFirmwareVersionHigher(currentVersion: String, availableVersion: String): Boolean {
        return PolarRuntimePlannerAdapter.isAvailableFirmwareVersionHigher(currentVersion, availableVersion)
    }

    suspend fun checkFirmwareUrlAvailability(
        client: BlePsFtpClient,
        identifier: String,
        firmwareUpdateApi: FirmwareUpdateApi
    ): Triple<String?, String?, FirmwareUpdateStatus> {
        val deviceInfo = readDeviceFirmwareInfo(client, identifier)
        val request = FirmwareUpdateRequest(
            clientId = "polar-sensor-data-collector-android",
            uuid = PolarDeviceUuid.fromDeviceId(identifier),
            firmwareVersion = deviceInfo.deviceFwVersion,
            hardwareCode = deviceInfo.deviceHardwareCode
        )
        val response = firmwareUpdateApi.checkFirmwareUpdate(request)
        return when (response.code()) {
            HttpResponseCodes.OK -> {
                val firmwareUpdateResponse = response.body()
                BleLogger.d(TAG, "Received firmware update response: $firmwareUpdateResponse")
                if (firmwareUpdateResponse != null &&
                    PolarRuntimePlannerAdapter.firmwareUpdateIsAvailable(deviceInfo.deviceFwVersion, firmwareUpdateResponse.version, firmwareUpdateResponse.fileUrl)) {
                    val availablePlan = PolarRuntimePlannerAdapter.planFirmwareCheckUpdateAvailableWorkflow()
                    require(availablePlan.statuses.last() == "checkFwUpdateAvailable")
                    Triple(firmwareUpdateResponse.version, firmwareUpdateResponse.fileUrl, FirmwareUpdateStatus.FetchingFwUpdatePackage("Firmware available, fetching"))
                } else {
                    val notAvailablePlan = PolarRuntimePlannerAdapter.planFirmwareCheckUpdateNotAvailableWorkflow()
                    require(notAvailablePlan.statuses.last() == "checkFwUpdateNotAvailable")
                    Triple(null, null, FirmwareUpdateStatus.FwUpdateNotAvailable("No fw update available, device firmware version ${deviceInfo.deviceFwVersion}"))
                }
            }
            HttpResponseCodes.NO_CONTENT -> {
                val notAvailablePlan = PolarRuntimePlannerAdapter.planFirmwareCheckUpdateNotAvailableWorkflow()
                require(notAvailablePlan.statuses.last() == "checkFwUpdateNotAvailable")
                Triple(null, null, FirmwareUpdateStatus.FwUpdateNotAvailable("No firmware update available"))
            }
            HttpResponseCodes.BAD_REQUEST -> {
                val errorBody = try { response.errorBody()?.string() ?: "Failed to read error body" } catch (e: Exception) { "Error reading error body: ${e.message}" }
                BleLogger.e("TAG", "Bad request to firmware update API: $errorBody")
                val clientFailurePlan = PolarRuntimePlannerAdapter.planFirmwareWorkflow(
                    id = "client-request-failure",
                    statuses = listOf("fwUpdateFailed")
                )
                require(clientFailurePlan.terminalError == "client-request-failure")
                Triple(null, null, FirmwareUpdateStatus.FwUpdateFailed("Bad request to firmware update API: $errorBody"))
            }
            else -> {
                val failedStatus = FirmwareUpdateStatus.FwUpdateFailed("Unexpected response code: ${response.code()}")
                if (isRetryableFirmwareAvailabilityFailure(failedStatus)) {
                    val retryableFailurePlan = PolarRuntimePlannerAdapter.planFirmwareRetryableServerFailureWorkflow()
                    require(retryableFailurePlan.terminalError == "retryable-server-failure")
                }
                Triple(null, null, failedStatus)
            }
        }
    }

    fun isRetryableFirmwareAvailabilityFailure(status: FirmwareUpdateStatus): Boolean {
        if (status !is FirmwareUpdateStatus.FwUpdateFailed) {
            return false
        }
        return PolarRuntimePlannerAdapter.firmwareAvailabilityFailureIsRetryable(status.details)
    }

    fun unzipFirmwarePackage(zipBytes: ByteArray): ByteArray {
        extractFirmwarePackagePayloads(zipBytes).firstOrNull()?.let { return it.second }
        try {
            val byteArrayInputStream = ByteArrayInputStream(zipBytes)
            val zipInputStream = ZipInputStream(byteArrayInputStream)
            val byteArrayOutputStream = ByteArrayOutputStream()

            zipInputStream.nextEntry
            val buffer = ByteArray(BUFFER_SIZE)
            var length: Int

            while (zipInputStream.read(buffer).also { length = it } != -1) {
                byteArrayOutputStream.write(buffer, 0, length)
            }

            zipInputStream.closeEntry()
            zipInputStream.close()

            return byteArrayOutputStream.toByteArray()
        } catch (e: Exception) {
            BleLogger.e(TAG, "Failed to unzip firmware package: $e")
            throw e
        }
    }

    fun extractFirmwarePackagePayloads(zipBytes: ByteArray): List<Pair<String, ByteArray>> {
        val firmwareFiles = mutableListOf<Pair<String, ByteArray>>()
        val buffer = ByteArray(BUFFER_SIZE)
        ZipInputStream(ByteArrayInputStream(zipBytes)).use { zipInputStream ->
            while (true) {
                val entry = zipInputStream.nextEntry ?: break
                val entryFileName = entry.name
                if (!PolarRuntimePlannerAdapter.firmwarePackageEntryIsPayload(entryFileName)) {
                    BleLogger.d(TAG, "Skipping firmware package entry $entryFileName")
                    zipInputStream.closeEntry()
                    continue
                }

                val byteArrayOutputStream = ByteArrayOutputStream()
                var length: Int
                while (zipInputStream.read(buffer).also { length = it } != -1) {
                    byteArrayOutputStream.write(buffer, 0, length)
                }
                BleLogger.d(TAG, "Extracted firmware package payload: $entryFileName")
                firmwareFiles.add(entryFileName to byteArrayOutputStream.toByteArray())
                zipInputStream.closeEntry()
            }
        }

        val orderedFirmwareNames = PolarRuntimePlannerAdapter.firmwarePayloadFileNames(firmwareFiles.map { it.first })
        val remainingFirmwareFiles = firmwareFiles.toMutableList()
        return orderedFirmwareNames.mapNotNull { fileName ->
            val index = remainingFirmwareFiles.indexOfFirst { it.first == fileName }
            if (index >= 0) remainingFirmwareFiles.removeAt(index) else null
        }
    }

    fun firmwareWriteFailure(error: Throwable, fileName: String): Throwable? {
        val terminal = (error as? PftpResponseError)?.let { pftpError ->
            PolarRuntimePlannerAdapter.firmwareWriteTerminal(pftpError.error, fileName)
        } ?: return error
        return when (terminal) {
            "success-rebooting" -> {
                val plan = PolarRuntimePlannerAdapter.planFirmwareSystemUpdateRebootSuccessWorkflow(listOf(fileName))
                require(plan.statuses.last() == "fwUpdateCompletedSuccessfully")
                require(plan.terminalError == null)
                null
            }
            "battery-too-low" -> {
                val plan = PolarRuntimePlannerAdapter.planFirmwareBatteryTooLowTerminalWorkflow(listOf(fileName))
                require(plan.statuses.last() == "fwUpdateFailed")
                require(plan.terminalError == "battery-too-low")
                PolarBleSdkInternalException("Battery too low to perform firmware update")
            }
            else -> error
        }
    }

    private fun devicePbVersionToString(pbVersion: Structures.PbVersion): String =
        PolarRuntimePlannerAdapter.firmwareDeviceVersion(pbVersion.major, pbVersion.minor, pbVersion.patch)
}

File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/impl/utils/PolarFirmwareUpdateUtils.swift
//  Copyright © 2024 Polar. All rights reserved.

import Foundation
import ZIPFoundation

protocol FirmwarePackageExtracting {
    func unzipFirmwarePackage(zippedData: Data) -> [String: Data]?
}

final class ZipFirmwarePackageExtractor: FirmwarePackageExtracting {
    func unzipFirmwarePackage(zippedData: Data) -> [String: Data]? {
        let temporaryDirectory = FileManager.default.temporaryDirectory

        let zipFilePath = temporaryDirectory.appendingPathComponent(UUID().uuidString + ".zip")
        do {
            try zippedData.write(to: zipFilePath)

            let destinationURL = temporaryDirectory.appendingPathComponent(UUID().uuidString)

            try FileManager.default.unzipItem(at: zipFilePath, to: destinationURL)

            let contents = try FileManager.default.contentsOfDirectory(at: destinationURL, includingPropertiesForKeys: nil)
            guard !contents.isEmpty else {
                BleLogger.error("unzipFirmwarePackage() error: No files found in the extracted directory")
                return nil
            }
            var fileDataDictionary: [String: Data] = [:]
            for fileURL in contents {
                let fileName = fileURL.lastPathComponent
                guard PolarFirmwareUpdateUtils.firmwarePackageEntryIsPayload(fileName) else {
                    BleLogger.trace("Skipping file: \(fileName)")
                    continue
                }
                let decompressedData = try Data(contentsOf: fileURL)
                fileDataDictionary[fileName] = decompressedData
                BleLogger.trace("Extracted file: \(fileName) - Size: \(decompressedData.count) bytes")
            }

            try FileManager.default.removeItem(at: zipFilePath)
            try FileManager.default.removeItem(at: destinationURL)

            return fileDataDictionary
        } catch {
            BleLogger.error("Error during unzipFirmwarePackage(): \(error)")
            return nil
        }
    }
}

class PolarFirmwareUpdateUtils {
    static let FIRMWARE_UPDATE_FILE_PATH = PolarRuntimePlanner.firmwareSystemUpdateFilePath()
    static let DEVICE_FIRMWARE_INFO_PATH = PolarRuntimePlanner.firmwareDeviceInfoPath()
    static var packageExtractor: FirmwarePackageExtracting = ZipFirmwarePackageExtractor()

    internal class FwFileComparator {
        static func compare(_ file1: String, _ file2: String) -> ComparisonResult {
            let firstPriority = PolarRuntimePlanner.firmwareFilePriority(file1)
            let secondPriority = PolarRuntimePlanner.firmwareFilePriority(file2)
            if firstPriority < secondPriority { return .orderedAscending }
            if firstPriority > secondPriority { return .orderedDescending }
            return .orderedSame
        }
    }
    
    static func readDeviceFirmwareInfo(client: BlePsFtpClient, deviceId: String) async -> PolarFirmwareVersionInfo? {
        let plannedOperation = deviceFirmwareInfoOperation()
        let request = plannedOperation ?? (.get, DEVICE_FIRMWARE_INFO_PATH)
        do {
            try ensureDeviceFirmwareInfoReadPlan()
            let serializedBytes = try PolarRuntimePlanner.fileOperationBytes(request)
            let response = try await client.request(serializedBytes)
            let proto = try Data_PbDeviceInfo(serializedBytes: response as Data)
            return PolarFirmwareVersionInfo(
                deviceFwVersion: devicePbVersionToString(pbVersion: proto.deviceVersion),
                deviceModelName: proto.modelName,
                deviceHardwareCode: proto.hardwareCode
            )
        } catch {
            BleLogger.error("Failed to request device info: \(deviceId), error: \(error)")
            return nil
        }
    }

    private static func deviceFirmwareInfoOperation() -> (command: Protocol_PbPFtpOperation.Command, path: String)? {
        return PolarRuntimePlanner.fileFacadeOperation(id: "firmware-read-device-info", command: "GET", path: DEVICE_FIRMWARE_INFO_PATH)
    }

    private static func ensureDeviceFirmwareInfoReadPlan() throws {
        let terminal = PolarRuntimePlanner.fileFacade(id: "firmware-read-device-info", command: "GET", path: DEVICE_FIRMWARE_INFO_PATH)
        guard terminal == "success" || terminal == "platform-owned" else {
            throw NSError(domain: "PolarFirmwareUpdateUtils", code: -1, userInfo: [NSLocalizedDescriptionKey: "Firmware device-info planning failed: \(terminal)"])
        }
    }

    static func isAvailableFirmwareVersionHigher(currentVersion: String, availableVersion: String) -> Bool {
        return PolarRuntimePlanner.isFirmwareVersionHigher(currentVersion: currentVersion, availableVersion: availableVersion)
    }

    static func firmwarePackageEntryIsPayload(_ fileName: String) -> Bool {
        return PolarRuntimePlanner.firmwarePackageEntryIsPayload(fileName)
    }

    static func firmwareFileTriggersRebootWait(_ fileName: String) -> Bool {
        return PolarRuntimePlanner.firmwareFileTriggersRebootWait(fileName)
    }

    static func unzipFirmwarePackage(zippedData: Data) -> [String: Data]? {
        return packageExtractor.unzipFirmwarePackage(zippedData: zippedData)
    }

    static func firmwareWriteFailure(error: Error, fileName: String, mapBatteryTooLow: () -> Error, mapError: (Error) -> Error) -> Error? {
        let errorCode = (error as NSError).code
        guard isPftpErrorCode(errorCode) else {
            return mapError(error)
        }
        let terminal = PolarRuntimePlanner.firmwareWriteTerminal(errorCode: errorCode, fileName: fileName)
        if terminal == "success-rebooting" {
            let workflowTerminal = PolarRuntimePlanner.firmwareSystemUpdateRebootSuccessWorkflow(fileNames: [fileName])
            guard workflowTerminal == "success" || workflowTerminal == "platform-owned" else {
                return mapError(NSError(domain: "PolarFirmwareUpdateUtils", code: -1, userInfo: [NSLocalizedDescriptionKey: "Firmware system reboot planning failed: \(workflowTerminal)"]))
            }
            return nil
        }
        if terminal == "battery-too-low" {
            let workflowTerminal = PolarRuntimePlanner.firmwareBatteryTooLowTerminalWorkflow(fileNames: [fileName])
            guard workflowTerminal == "success" || workflowTerminal == "platform-owned" else {
                return mapError(NSError(domain: "PolarFirmwareUpdateUtils", code: -1, userInfo: [NSLocalizedDescriptionKey: "Firmware battery terminal planning failed: \(workflowTerminal)"]))
            }
            if let terminalError = PolarRuntimePlanner.firmwareBatteryTooLowTerminalError(fileNames: [fileName]), terminalError != "battery-too-low" {
                return mapError(NSError(domain: "PolarFirmwareUpdateUtils", code: -1, userInfo: [NSLocalizedDescriptionKey: "Firmware battery terminal-error planning failed: \(terminalError)"]))
            }
            return mapBatteryTooLow()
        }
        return mapError(error)
    }

    private static func isPftpErrorCode(_ code: Int) -> Bool {
        return code == 0 || code == 1 || code == 2 || (100...108).contains(code) || (200...209).contains(code)
    }
    
    private static func devicePbVersionToString(pbVersion: PbVersion) -> String {
        return PolarRuntimePlanner.firmwareDeviceVersion(major: Int(pbVersion.major), minor: Int(pbVersion.minor), patch: Int(pbVersion.patch))
    }
}
